File: backend/app/models/presenca.py
from datetime import date
from enum import Enum

from pydantic import BaseModel


class StatusPresenca(str, Enum):
    PRESENTE = "presente"
    FALTA = "falta"
    FALTA_JUSTIFICADA = "falta_justificada"


class PresencaItem(BaseModel):
    catequisando_id: str
    status: StatusPresenca


class MarcarPresencasRequest(BaseModel):
    fase_id: str
    data: date
    presencas: list[PresencaItem]


class PresencaOut(BaseModel):
    catequisando_id: str
    catequisando_nome: str
    status: StatusPresenca


class ListaPresencasOut(BaseModel):
    fase_id: str
    data: date
    presencas: list[PresencaOut]


class RegistoPresenca(BaseModel):
    data: date
    status: StatusPresenca


class HistoricoPresencasOut(BaseModel):
    catequisando_id: str
    total_registos: int
    total_presencas: int
    total_faltas: int
    total_faltas_justificadas: int
    registos: list[RegistoPresenca]


class LinhaRelatorioPresencas(BaseModel):
    catequisando_id: str
    nome: str
    presencas: int
    faltas: int
    faltas_justificadas: int
    total: int

File: frontend/lib/models/presenca.dart
/// Os 3 estados possíveis de presença numa catequese.
enum StatusPresenca {
  presente,
  falta,
  faltaJustificada;

  String get valor {
    switch (this) {
      case StatusPresenca.presente:
        return 'presente';
      case StatusPresenca.falta:
        return 'falta';
      case StatusPresenca.faltaJustificada:
        return 'falta_justificada';
    }
  }

  String get rotulo {
    switch (this) {
      case StatusPresenca.presente:
        return 'Presença';
      case StatusPresenca.falta:
        return 'Falta';
      case StatusPresenca.faltaJustificada:
        return 'Falta Justificada';
    }
  }

  String get rotuloCurto {
    switch (this) {
      case StatusPresenca.presente:
        return 'P';
      case StatusPresenca.falta:
        return 'F';
      case StatusPresenca.faltaJustificada:
        return 'FJ';
    }
  }

  static StatusPresenca fromValor(String valor) {
    switch (valor) {
      case 'presente':
        return StatusPresenca.presente;
      case 'falta_justificada':
        return StatusPresenca.faltaJustificada;
      case 'falta':
      default:
        return StatusPresenca.falta;
    }
  }
}

class RegistoPresenca {
  final DateTime data;
  final StatusPresenca status;

  RegistoPresenca({required this.data, required this.status});

  factory RegistoPresenca.fromJson(Map<String, dynamic> json) => RegistoPresenca(
        data: DateTime.parse(json['data'] as String),
        status: StatusPresenca.fromValor(json['status'] as String),
      );
}

class HistoricoPresencas {
  final int totalRegistos;
  final int totalPresencas;
  final int totalFaltas;
  final int totalFaltasJustificadas;
  final List<RegistoPresenca> registos;

  HistoricoPresencas({
    required this.totalRegistos,
    required this.totalPresencas,
    required this.totalFaltas,
    required this.totalFaltasJustificadas,
    required this.registos,
  });

  factory HistoricoPresencas.fromJson(Map<String, dynamic> json) => HistoricoPresencas(
        totalRegistos: json['total_registos'] as int,
        totalPresencas: json['total_presencas'] as int,
        totalFaltas: json['total_faltas'] as int,
        totalFaltasJustificadas: json['total_faltas_justificadas'] as int,
        registos: (json['registos'] as List? ?? [])
            .map((e) => RegistoPresenca.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class PresencaItem {
  final String catequisandoId;
  final String catequisandoNome;
  StatusPresenca status;

  PresencaItem({
    required this.catequisandoId,
    required this.catequisandoNome,
    required this.status,
  });

  factory PresencaItem.fromJson(Map<String, dynamic> json) => PresencaItem(
        catequisandoId: json['catequisando_id'] as String,
        catequisandoNome: json['catequisando_nome'] as String,
        status: StatusPresenca.fromValor(json['status'] as String),
      );

  Map<String, dynamic> toRequestJson() => {
        'catequisando_id': catequisandoId,
        'status': status.valor,
      };
}
